apps/documents/serializers.py

```diff
@@ -80,11 +84,20 @@ def get_presigned_s3(self, obj):
         if obj.upload_status != UploadStatus.COMPLETE:
             return super().get_presigned_s3(obj)
 
+        try:
+            settings.S3_CLIENT.head_object(Bucket=settings.S3_BUCKET, Key=obj.s3_key)
+        except ClientError:
+            # The document doesn't exist anymore in the bucket. The bucket is going to be repopulated from vault
+            storage_credentials_id, wallet_id, vault_id, filename = obj.s3_key.split('/', 3)
+
+            DocumentRPCClient().put_document_in_s3(settings.S3_BUCKET, obj.s3_key, wallet_id, storage_credentials_id,
```

Comment:
One last change. In the Document model, make the following change:
```
    @property
    def filename(self):
        return f"{self.id}.{self.file_type.name.lower()}"

    @property
    def s3_key(self):
        return f"{self.shipment.storage_credentials_id}/{self.shipment.shipper_wallet_id}" \
            f"/{self.shipment.vault_id}/{self.filename}"
```

And then on this line you can do
```suggestion
            DocumentRPCClient().put_document_in_s3(settings.S3_BUCKET, obj.s3_key, obj.shipment.shipper_wallet_id, obj.shipment.storage_credentials_id,
```

Or if you think it's cleaner, you can add additional Document properties for storage_credentials_id, wallet_id, etc.